Add SO_REUSEPORT support for EntryPoints

aofei opened the pull request:
### What does this PR do?

Add SO_REUSEPORT support for EntryPoints. (Fixes #9823)

### Motivation

Attempts to solve the problem of canary deployment against Traefik itself.

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

This PR only focuses on the implementation. For discussion, please move to #9823.

mloiseleur commented:
I tested this PR.
It works like a charm.

rtribotte commented:
Hello @aofei,

I can see that the update of the static configuration references files is missing:
- `docs/content/reference/static-configuration/file.toml`
- `docs/content/reference/static-configuration/file.yaml`

Can you please update them?

aofei commented:
Hi @rtribotte!

I have just updated those two files and rebased my branch as well.

aofei commented:
Also, should I change the base branch of this PR from `master` to `v3.0`?

rtribotte commented:
@aofei Thanks for the changes!

> Also, should I change the base branch of this PR from master to v3.0?

For now, I don't know, I'm going to raise a discussion with other maintainers on that subject and let you know.

rtribotte commented:
Hello @aofei,

I came across this [article](https://lwn.net/Articles/853637/) mentioning limitations associated with the SO_REUSEPORT directive, especially the closing of connections by the kernel when one of the processes using the port exits:

> There are a number of reasons why a listening process might exit on a busy system. Perhaps it simply crashed. But, more likely, the server is being restarted to effect a configuration change or to switch to a new certificate. Such restarts can be phased across a pool of server processes so that they don't all exit at once; that should allow incoming connections to be handled without any apparent interruption of service. But when the above-described behavior comes into the picture, users can be turned away, which tends to have an unpleasant effect on their mood. The depressive effect on the operator of the site, who may have just lost the opportunity to learn that the would-be user is in the market for a new pair of socks, can be even worse.

Since you implemented this feature, I'm hoping to get your insights, but so far, these limitations could be a significant concern for Traefik users, at least I think this should be documented.

rtribotte commented:
> Also, should I change the base branch of this PR from master to v3.0?

So, as a follow-up, yes please, if you can rebase this PR on the v3.0 branch it would be great, we would like to have it for v3.

aofei commented:
@rtribotte Yes, you're right, thanks for bringing it up! It's a known bug in the SO_REUSEPORT implementation of the Linux kernel.

In case anyone else is interested, the problem mentioned by @rtribotte is that when a listening process exits, all incoming TCP connections assigned to it that are either in the middle of a 3WH (3-way handshake) or in the accept queue get closed, even if other processes could handle them. The root cause of this problem is that once the kernel assigns an incoming connection to a specific process (triggered by the arrival of the initial `SYN` packet of a 3WH), it won't reassign it, leading to connection failures (`RST` packets) when that process exits.

A common workaround is to pass the listener file descriptors between different processes without closing them. For example, Nginx does this, thanks to its concept of worker processes. Unfortunately, this doesn't work with Traefik, as Traefik is a single process.

While this is a problem, it doesn't make SO_REUSEPORT a bad feature. It's a kernel bug, until it's resolved or a workaround is found, I agree that it should be documented.

aofei commented:
@rtribotte I have updated the `docs/content/routing/entrypoints.md` documentation file and changed the base branch of this PR from master to v3.0. The failed checks don't seem relevant to this PR. Please let me know if I've got something wrong.

nmengin commented:
Hey @aofei,

It seems that you have conflicts, could you fix them?
Do you need help on it?

aofei commented:
Hi @nmengin! Thanks for letting me know about the conflicts. They occurred after my last commit, so I didn't notice. They have now been fixed.

nmengin commented:
Hey @aofei,

Thank you for your quick answer and action.

kevinpollet commented on pkg/udp/conn.go:
```diff
@@ -33,18 +34,18 @@ type Listener struct {
-	conn, err := net.ListenUDP(network, laddr)
+	packetConn, err := listenConfig.ListenPacket(context.Background(), network, address)
 	if err != nil {
 		return nil, err
```
```suggestion
		return nil, fmt.Errorf("listen packet: %w", err)
```

kevinpollet reviewed (changes requested):
Thanks for this contribution!

aofei commented on pkg/config/static/entrypoint_getlistenconfig_unix.go:
```diff
@@ -0,0 +1,36 @@
+				err = unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unixSOREUSEPORT, 1)
+				if err != nil {
+					return
+				}
```
I recommend reading https://stackoverflow.com/a/14388707/5801959, it is a very detailed answer.

To put it simply, due to various historical reasons and differences in how different kernels implement them, we must set both of them at the same time to get what we want.

rtribotte replied:
@aofei Thanks for this explanation.
However, I'm wondering why we couldn't make only use of `SO_REUSEPORT`.
What was, regarding the `reusePort` feature in Traefik, the rationale for using `SO_REUSEADDR`?
If it was for <3.9 kernels support, the `SetsockoptInt` call with `SO_REUSEPORT` would likely return an error, so I guess the option would not work well with old kernels anyway.

aofei replied:
Setting `SO_REUSEPORT` without `SO_REUSEADDR` may lead to issues like inability to immediately rebind a server to a port if the previous instance left sockets in the `TIME_WAIT` state.

And from https://man7.org/linux/man-pages/man7/socket.7.html:

> SO_REUSEPORT (since Linux 3.9)
>
> Permits multiple AF_INET or AF_INET6 sockets to be bound
> to an identical socket address.
> …

If only `SO_REUSEPORT` is set, we cannot bind two sockets to `0.0.0.0:80` and `192.168.1.2:80`, respectively, because `SO_REUSEPORT` typically requires all sockets to bind to an identical socket address. However, this behavior can vary with different kernel implementations. For example, Darwin does not impose such a restriction.

---

Despite the above issues, it appears we can safely omit setting `SO_REUSEADDR` in Traefik, as `net.ListenConfig` already [sets it by default](https://github.com/golang/go/blob/923ab13f9bb6fccc362d5259e2f456abbeda0c51/src/net/sockopt_linux.go#L26-L35). Sorry I just found out about this today.

So, I think we can remove the code that sets `SO_REUSEADDR` and leave a comment referencing https://go.dev/src/net/sockopt_linux.go. What do you think @rtribotte?

rtribotte replied:
@aofei Good point indeed for the `TIME_WAIT` state.
I'm not sure about the wildcard IPs by looking at the output of the [code](https://rextester.com/BUAFK86204) shared in the Stackoverflow post.
It looks like when `SO_REUSEPORT` is set on both sockets, it does not require all sockets to bind to an identical socket address.
But it is surely depending on the kernel implementation.

Yes, since `net.ListenConfig` does already set it by default, and does it accordingly to the OS, I think it is better to only set `SO_REUSEPORT`.

Thanks!

rtribotte replied:
I'm marking the PR as waiting for corrections.

aofei commented on pkg/udp/conn.go:
```diff
@@ -33,18 +34,18 @@ type Listener struct {
 
 	l := &Listener{
-		pConn:     conn,
+		pConn:     packetConn.(*net.UDPConn),
```
Actually, `packetConn` is always a `*net.UDPConn` in this case, so there's no risk of panic. But I've made the change you requested, since adding the nil check could provide extra reassurance.

nmengin reviewed (approved):
LGTM 👏 

Thank you @aofei 👍

kevinpollet commented on pkg/config/static/entrypoint_listenconfig_other.go:
Hello @aofei,

Thanks for making the changes, 

After discussing with other maintainers, we think that those files have to moved to the `server` pkg and provide a function to build a `ListenConfig` in function of a given EntryPoints configuration.

Please could you move those files and rename them to `server_entrypoint_listenconfig_*`?

aofei commented:
@kevinpollet @rtribotte I just made the changes you requested. Please let me know if I've got something wrong.

kevinpollet reviewed (approved):
Thanks 👍

rtribotte reviewed (approved):
Thanks!

juliens reviewed (approved):
LGTM

nmengin reviewed (approved):
LGTM
